src/shared/event-emitter.ts

```diff
@@ -0,0 +1,49 @@
+export type EventMap = Record<string, (...args: any) => void>;
+
+/**
+ * A simple event emitter with an API similar to Node's `EventEmitter`.
+ *
+ * `Event` is an object type that maps event names to the function signatures of
+ * subscribers for those events. For example, this defines an emitter which
+ * emits one event, `uriChanged`, with a single string argument:
+ *
+ * ```
+ * type Events = {
+ *   uriChanged(uri: string): void;
+ * }
+ * ```
+ */
+export class EventEmitter<Event extends EventMap> {
+  // Use a private field here to avoid conflicts with subclasses.
+  #listeners: Array<{ name: keyof Event; callback: (...args: any) => void }> =
```

Comment:
Are we going to move to this approach for private fields rather than using the `private` keyword and prefixing with underscore?

I would be in favor of that.

Reply:
Support in JS engines and tools is now good enough that I think we could consider it. There might be a handful of places in the client where tests still access internal properties.